src/script/interpreter.cpp

```diff
@@ -2092,6 +2091,105 @@ bool VerifyTaprootCommitment(uint256 &tapleaf_hash,
     return q == q_expected;
 }
 
+#include <logging.h>
+#include <util/strencodings.h>
+
+static bool VerifyScriptType(std::vector<valtype> stack,
+                             const CScript &script_pubkey, uint32_t flags,
+                             const BaseSignatureChecker &checker,
+                             ScriptExecutionMetrics &metrics_out,
+                             ScriptError *serror) {
+    if (script_pubkey.size() == 1 ||
+        !script_pubkey.IsPushOnly(script_pubkey.begin() + 1)) {
+        return set_error(serror, ScriptError::SCRIPTTYPE_MALFORMED_SCRIPT);
+    }
+    if (script_pubkey[1] == OP_1) { // Taproot script version
+        if (!((script_pubkey.size() == SCRIPT_SIZE_WITHOUT_STATE &&
+               script_pubkey[SCRIPT_INTRO_SIZE - 1] ==
+                   CPubKey::COMPRESSED_SIZE) ||
+              (script_pubkey.size() == SCRIPT_SIZE_WITH_STATE &&
+               script_pubkey[SCRIPT_INTRO_SIZE + CPubKey::COMPRESSED_SIZE] ==
+                   CSHA256::OUTPUT_SIZE))) {
+            return set_error(serror, ScriptError::SCRIPTTYPE_MALFORMED_SCRIPT);
+        }
+        valtype vch_pubkey =
+            valtype(script_pubkey.begin() + SCRIPT_INTRO_SIZE,
+                    script_pubkey.begin() + SCRIPT_SIZE_WITHOUT_STATE);
+
+        if (stack.size() == 0) {
+            return set_error(serror, ScriptError::INVALID_STACK_OPERATION);
+        }
+        if (stack.size() >= 2 && !stack.back().empty() &&
+            stack.back()[0] == ANNEX_TAG) {
+            return set_error(serror, ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED);
+        }
+        if (stack.size() == 1) {
+            // Spend using single signature instead of executing script
+            valtype &vch_sig = stacktop(-1);
+            bool f_success = false;
+            if (!EvalChecksig(vch_sig, vch_pubkey, script_pubkey.begin(),
+                              script_pubkey.end(),
+                              flags | SCRIPT_REQUIRE_TAPROOT_SIGHASH, checker,
+                              metrics_out, std::nullopt, serror, f_success)) {
+                if (*serror == ScriptError::SIG_NULLFAIL) {
+                    *serror = ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED;
+                }
+                // serror is set
+                return false;
+            }
+            if (!f_success) {
+                return set_error(serror,
+                                 ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED);
+            }
+            return set_success(serror);
+        } else {
```

Comment:
The above branch always returns. Please remove this else just to make it easier to read this below stuff.